# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** Alejandro De Santiago Palomares Salinas

**Student ID:** aeb923

**Date Submitted:** 10/17/2025

**Due Date:** 10/12/2025

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [20]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)
setwd("/workspaces/assignment-2-version3-Aledesan-utsa/data")
getwd()

# TODO: Load lubridate
library(lubridate)

cat("✅ Packages loaded successfully!\n")

[1] "/workspaces/assignment-2-version3-Aledesan-utsa/data"

✅ Packages loaded successfully!


In [21]:
# Task 1.2: Import Datasets
# TODO: Import customer_feedback.csv into a variable called 'feedback'
feedback <- read.csv("customer_feedback.csv")

# TODO: Import transaction_log.csv into a variable called 'transactions'
transactions <- read.csv("transaction_log.csv")

# TODO: Import product_catalog.csv into a variable called 'products'
products <- read.csv("product_catalog.csv")

cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n")

✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 150 
Product rows: 75 


In [22]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# TODO: Display structure of feedback using str()
str(feedback)

# TODO: Display first 5 rows of feedback
head(feedback,5)

cat("\n=== TRANSACTION DATA ===\n")
# TODO: Display structure of transactions
str(transactions)

# TODO: Display first 5 rows of transactions
head(transactions,5)

cat("\n=== PRODUCT CATALOG DATA ===\n")
# TODO: Display structure of products
str(products)

# TODO: Display first 5 rows of products
head(products,5)


=== CUSTOMER FEEDBACK DATA ===
'data.frame':	100 obs. of  6 variables:
 $ FeedbackID   : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : int  12 40 34 1 47 13 13 37 49 23 ...
 $ Customer_Name: chr  "Bob Wilson" "   sarah johnson   " "jane smith" "JANE SMITH" ...
 $ Feedback_Text: chr  "Highly recommend this item" "   Excellent service   " "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr  "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: chr  "2024-02-23" "2024-01-21" "2023-09-02" "2023-08-21" ...


FeedbackID CustomerID Customer_Name       Feedback_Text                   
1 1          12         Bob Wilson          Highly recommend this item      
2 2          40            sarah johnson       Excellent service            
3 3          34         jane smith          Poor quality control            
4 4           1         JANE SMITH          average product, nothing special
5 5          47         michael brown       AMAZING customer support!!!     
  Contact_Info           Feedback_Date
1 bob.wilson@test.org    2024-02-23   
2 555-123-4567           2024-01-21   
3 jane_smith@company.com 2023-09-02   
4 jane_smith@company.com 2023-08-21   
5 555-123-4567           2023-04-24


=== TRANSACTION DATA ===
'data.frame':	150 obs. of  5 variables:
 $ LogID               : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID          : int  26 21 12 6 32 27 31 30 31 13 ...
 $ Transaction_DateTime: chr  "4/5/24 14:30" "3/15/24 14:30" "3/15/24 14:30" "3/20/24 9:15" ...
 $ Amount              : num  277 175 252 215 269 ...
 $ Status              : chr  "Pending" "Pending" "Pending" "Pending" ...


LogID CustomerID Transaction_DateTime Amount Status   
1 1     26         4/5/24 14:30         277.22 Pending  
2 2     21         3/15/24 14:30        175.16 Pending  
3 3     12         3/15/24 14:30        251.71 Pending  
4 4      6         3/20/24 9:15         214.98 Pending  
5 5     32         3/20/24 9:15         268.91 Completed


=== PRODUCT CATALOG DATA ===
'data.frame':	75 obs. of  5 variables:
 $ ProductID          : int  1 2 3 4 5 6 7 8 9 10 ...
 $ Product_Description: chr  "Apple iPhone 14 Pro - 128GB - Space Black" "samsung galaxy s23 ultra 256gb" "Apple iPhone 14 Pro - 128GB - Space Black" "Apple iPhone 14 Pro - 128GB - Space Black" ...
 $ Category           : chr  "TV" "TV" "Audio" "Shoes" ...
 $ Price              : num  964 1817 853 649 586 ...
 $ In_Stock           : chr  "Limited" "Yes" "Yes" "Yes" ...


ProductID Product_Description                       Category    Price  
1 1         Apple iPhone 14 Pro - 128GB - Space Black TV           963.53
2 2         samsung galaxy s23 ultra 256gb            TV          1817.44
3 3         Apple iPhone 14 Pro - 128GB - Space Black Audio        852.79
4 4         Apple iPhone 14 Pro - 128GB - Space Black Shoes        648.58
5 5         samsung galaxy s23 ultra 256gb            Electronics  586.35
  In_Stock
1 Limited 
2 Yes     
3 Yes     
4 Yes     
5 Limited

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [23]:
# Task 2.1: Clean Product Names
# TODO: Create a new column 'product_name_clean' that:
#   - Removes leading/trailing whitespace using str_trim()
#   - Converts to Title Case using str_to_title()

products_name_clean <- products %>%
  mutate(
    product_name_clean = str_trim(Product_Description),
    product_name_clean = str_to_title(Product_Description),
  )

# Display before and after
cat("Product Name Cleaning Results:\n")
products_name_clean %>%
  select(product_name_clean) %>%
  head(10) %>%
  print()

Product Name Cleaning Results:
                          product_name_clean
1  Apple Iphone 14 Pro - 128gb - Space Black
2             Samsung Galaxy S23 Ultra 256gb
3  Apple Iphone 14 Pro - 128gb - Space Black
4  Apple Iphone 14 Pro - 128gb - Space Black
5             Samsung Galaxy S23 Ultra 256gb
6  Apple Iphone 14 Pro - 128gb - Space Black
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram
8         Hp Envy Printer - Wireless - Color
9   Nike Air Max 270 - Size 10 - Black/White
10         Lg 55" 4k Smart Tv - Oled Display


In [24]:
# Task 2.2: Standardize Product Categories
# TODO: Create a new column 'category_clean' that:
#   - Converts category to Title Case
#   - Removes any extra whitespace

products_clean <- products %>%
  mutate(
   category_clean = str_to_title(Category)   
  )

# Show unique categories before and after
cat("Original categories:\n")
print(unique(products$category))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))

Original categories:
NULL

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [25]:
# Task 2.3: Clean Customer Feedback Text
# TODO: Create a new column 'feedback_clean' that:
#   - Converts text to lowercase using str_to_lower()
#   - Removes extra whitespace using str_squish()

feedback_clean <- feedback %>%
  mutate(
    feedback_clean = str_to_title(Feedback_Text) 
    
  )

# Display sample
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
  select(Feedback_Text, feedback_clean) %>%
  head(5) %>%
  print()

Feedback Cleaning Sample:
                     Feedback_Text                   feedback_clean
1       Highly recommend this item       Highly Recommend This Item
2             Excellent service                Excellent Service   
3             Poor quality control             Poor Quality Control
4 average product, nothing special Average Product, Nothing Special
5      AMAZING customer support!!!      Amazing Customer Support!!!


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [26]:
# Task 3.1: Detect Product Features
# TODO: Create three new columns:
#   - is_wireless: TRUE if product name contains "wireless" (case-insensitive)
#   - is_premium: TRUE if product name contains "pro", "premium", or "deluxe"
#   - is_gaming: TRUE if product name contains "gaming" or "gamer"
# Hint: Use str_detect() with str_to_lower() for case-insensitive matching
# Hint: Use | (pipe) in regex for OR conditions

products_clean <- products %>%
  mutate(
    product_name_clean = str_to_title(Product_Description),  # First create clean product name
    is_wireless = str_detect(str_to_lower(product_name_clean), "wireless"),
    is_gaming = str_detect(str_to_lower(product_name_clean), "gaming|gamer"),
    is_premium = str_detect(str_to_lower(product_name_clean), "pro|premium|deluxe")
  )    

# Display results
cat("Product Feature Detection:\n")
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# Summary statistics
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless), "\n")
cat("Premium products:", sum(products_clean$is_premium), "\n")
cat("Gaming products:", sum(products_clean$is_gaming), "\n")

Product Feature Detection:
                          product_name_clean is_wireless is_premium is_gaming
1  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
2             Samsung Galaxy S23 Ultra 256gb       FALSE      FALSE     FALSE
3  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
4  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
5             Samsung Galaxy S23 Ultra 256gb       FALSE      FALSE     FALSE
6  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram       FALSE      FALSE     FALSE
8         Hp Envy Printer - Wireless - Color        TRUE      FALSE     FALSE
9   Nike Air Max 270 - Size 10 - Black/White       FALSE      FALSE     FALSE
10         Lg 55" 4k Smart Tv - Oled Display       FALSE      FALSE     FALSE

Feature Summary:
Wireless products: 17 
Premium products: 13 
Gaming products: 0 


In [27]:
# Task 3.2: Extract Product Specifications
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
   size_number = str_extract(product_name_clean, "\\d+"),
  )

# Display products with extracted sizes
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()

Extracted Product Specifications:
                          product_name_clean size_number
1  Apple Iphone 14 Pro - 128gb - Space Black          14
2             Samsung Galaxy S23 Ultra 256gb          23
3  Apple Iphone 14 Pro - 128gb - Space Black          14
4  Apple Iphone 14 Pro - 128gb - Space Black          14
5             Samsung Galaxy S23 Ultra 256gb          23
6  Apple Iphone 14 Pro - 128gb - Space Black          14
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram          13
8   Nike Air Max 270 - Size 10 - Black/White         270
9          Lg 55" 4k Smart Tv - Oled Display          55
10       Sony Wh-1000xm4 Wireless Headphones        1000


In [28]:
# Task 3.3: Simple Sentiment Analysis
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
    positive_words = str_count(str_to_lower(feedback_clean), "great|excellent|love|amazing|highly|good"),
    negative_words = str_count(str_to_lower(feedback_clean), "poor|bad|terrible|hate|awful"),
    sentiment_score = positive_words - negative_words
  )

# Display sentiment analysis results
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# Summary
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0), "\n")

Sentiment Analysis Results:
                     feedback_clean positive_words negative_words
1        Highly Recommend This Item              1              0
2              Excellent Service                 1              0
3              Poor Quality Control              0              1
4  Average Product, Nothing Special              0              0
5       Amazing Customer Support!!!              1              0
6       Amazing Customer Support!!!              1              0
7  Average Product, Nothing Special              0              0
8              Good Value For Money              1              0
9        Highly Recommend This Item              1              0
10       Highly Recommend This Item              1              0
   sentiment_score
1                1
2                1
3               -1
4                0
5                1
6                1
7                0
8                1
9                1
10               1

Overall Sentiment Summary:
Average s

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [29]:
# Task 4.1: Parse Transaction Dates
# TODO: Create a new column 'date_parsed' that parses the transaction_date column
# Hint: Check the format of transaction_date first, then use ymd(), mdy(), or dmy()

transactions_clean <- transactions %>%
  mutate(
    date_parsed = mdy_hm(Transaction_DateTime),
    date_parsed = if_else(is.na(date_parsed), 
                          dmy_hms(Transaction_DateTime), 
                          date_parsed),
    date_only = as_date(date_parsed),
  )

# Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(date_only) %>%
  head(10) %>%
  print()

Warning message:
“There were 2 warnings in `mutate()`.
The first warning was:
ℹ In argument: `date_parsed = mdy_hm(Transaction_DateTime)`.
Caused by warning:
!  61 failed to parse.
ℹ Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.”


Date Parsing Results:
    date_only
1  2024-04-05
2  2024-03-15
3  2024-03-15
4  2024-03-20
5  2024-03-20
6  2024-03-20
7  2024-03-20
8  2024-03-15
9  2024-03-25
10 2024-04-05


In [30]:
# Task 4.2: Extract Date Components
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
    trans_year = year(date_only),
    trans_month = month(date_only),
    trans_month_name = month(date_only, label = TRUE, abbr = FALSE),
    trans_day = day(date_only),
    trans_weekday = wday(date_only, label = TRUE, abbr = FALSE),
    trans_quarter = quarter(date_only)
  )

# Display results
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_only, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()

Date Component Extraction:
    date_only trans_month_name trans_weekday trans_quarter
1  2024-04-05            April        Friday             2
2  2024-03-15            March        Friday             1
3  2024-03-15            March        Friday             1
4  2024-03-20            March     Wednesday             1
5  2024-03-20            March     Wednesday             1
6  2024-03-20            March     Wednesday             1
7  2024-03-20            March     Wednesday             1
8  2024-03-15            March        Friday             1
9  2024-03-25            March        Monday             1
10 2024-04-05            April        Friday             2


In [31]:
# Task 4.3: Identify Weekend Transactions
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
    is_weekend = wday(date_parsed) %in% c(1, 7)
  )

# Summary
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")

Weekend vs Weekday Transactions:

FALSE 
  150 

Percentage of weekend transactions: 0 %


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [32]:
# Task 5.1: Calculate Days Since Transaction
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
left_join(feedback %>% select(CustomerID, Customer_Name), by = "CustomerID") %>%
  mutate(
    days_since = as.numeric(today() - as.Date(date_parsed)),
  )
#used left join bc i couldnt figure out another way to get customer names to display here
# Display results
cat("Days Since Transaction:\n")
transactions_clean %>% 
  select(Customer_Name, date_only, days_since) %>%
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()

Warning message in left_join(., feedback %>% select(CustomerID, Customer_Name), :
“Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 1 of `x` matches multiple rows in `y`.
ℹ Row 76 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  "many-to-many"` to silence this warning.”


Days Since Transaction:
   Customer_Name  date_only days_since
1  mary thompson 2024-03-15        581
2  MARY THOMPSON 2024-03-15        581
3     Bob Wilson 2024-03-15        581
4     bob wilson 2024-03-15        581
5     BOB WILSON 2024-03-15        581
6     bob wilson 2024-03-15        581
7    william kim 2024-03-15        581
8    william kim 2024-03-15        581
9  michael brown 2024-03-15        581
10 michael brown 2024-03-15        581


In [33]:
# Task 5.2: Categorize by Recency
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
    recency_category = case_when(
      days_since <= 30 ~ "Recent (< 30 days)",
      days_since <= 90 ~ "Moderate (30-90 days)",
      TRUE ~ "At Risk (> 90 days)"
  ))

# Display distribution
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# Show at-risk customers
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(Customer_Name, date_only, days_since) %>%
  arrange(desc(days_since)) %>%
  print()

Recency Category Distribution:

At Risk (> 90 days) 
                313 

At-Risk Customers (>90 days):
[1] Customer_Name date_only     days_since   
<0 rows> (or 0-length row.names)


## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [34]:
# Task 6.1: Extract First Names and Create Personalized Messages
# TODO: Create two new columns:
#   - first_name: Extract first name from customer_name (everything before first space)
#   - personalized_message: Create message based on recency_category
#     * Recent: "Hi [name]! Thanks for your recent purchase!"
#     * Moderate: "Hi [name], we miss you! Check out our new products."
#     * At Risk: "Hi [name], it's been a while! Here's a special offer for you."
# Hint: Use str_extract() with pattern "^\\\\w+" for first name
# Hint: Use paste() to combine strings in case_when()

customer_outreach <- transactions_clean %>%
  mutate(
    first_name = str_extract(Customer_Name, "^\\w+"),
    personalized_message = case_when(
  days_since <= 30 ~ paste("Hi", first_name, "! Thanks for your recent purchase!"),
  days_since <= 90 ~ paste("Hi", first_name, ", we miss you! Check out our new products."),
  TRUE ~ paste("Hi", first_name, ", it's been a while! Here's a special offer for you.")
    )
  )
# Display personalized messages
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(Customer_Name, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()

Personalized Customer Messages:
   Customer_Name first_name days_since
1   SUSAN HARRIS      SUSAN        560
2   SUSAN HARRIS      SUSAN        560
3   susan harris      susan        560
4  mary thompson       mary        581
5  MARY THOMPSON       MARY        581
6     Bob Wilson        Bob        581
7     bob wilson        bob        581
8     BOB WILSON        BOB        581
9     bob wilson        bob        581
10  CHRIS MARTIN      CHRIS        576
                                            personalized_message
1  Hi SUSAN , it's been a while! Here's a special offer for you.
2  Hi SUSAN , it's been a while! Here's a special offer for you.
3  Hi susan , it's been a while! Here's a special offer for you.
4   Hi mary , it's been a while! Here's a special offer for you.
5   Hi MARY , it's been a while! Here's a special offer for you.
6    Hi Bob , it's been a while! Here's a special offer for you.
7    Hi bob , it's been a while! Here's a special offer for you.
8    Hi BOB , it's 

In [35]:
# Task 6.2: Analyze Transaction Patterns by Weekday
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending

weekday_patterns <- transactions_clean %>%
  filter(!is.na(trans_weekday)) %>%
  group_by(trans_weekday) %>%
  summarize(
    transaction_count = n(),
    total_revenue = sum(Amount),
    avg_transaction = mean(Amount),
    .groups = 'drop'
  ) %>%
  arrange(desc(transaction_count))
  
# Display results
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# Identify busiest day
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")

Transaction Patterns by Weekday:
# A tibble: 3 × 4
  trans_weekday transaction_count total_revenue avg_transaction
  <ord>                     <int>         <dbl>           <dbl>
1 Friday                      122        34130.            280.
2 Wednesday                    59        13840.            235.
3 Monday                       58        14562.            251.

🔥 Busiest day: Friday 


In [36]:
# Task 6.3: Monthly Transaction Analysis
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
filter(!is.na(trans_month_name)) %>%
  group_by(trans_month_name) %>%
  summarize(
    Transaction_Count = n(),
    disrtinct_Customers = n_distinct(Customer_Name),
    .groups = 'drop'
  ) %>%
  arrange(trans_month_name)

print("Monthly Revenue Summary:")
print(monthly_patterns)
  
# Display results
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)

[1] "Monthly Revenue Summary:"
# A tibble: 2 × 3
  trans_month_name Transaction_Count disrtinct_Customers
  <ord>                        <int>               <int>
1 March                          174                  38
2 April                           65                  30
Monthly Transaction Patterns:
# A tibble: 2 × 3
  trans_month_name Transaction_Count disrtinct_Customers
  <ord>                        <int>               <int>
1 March                          174                  38
2 April                           65                  30


## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [39]:
# Task 7.1: Create Business Intelligence Dashboard

cat("\n", rep("=", 60), "\n")
cat("BUSINESS INTELLIGENCE SUMMARY\n")
cat(rep("=", 60), "\n\n")

# Product Analysis
cat("📦 PRODUCT ANALYSIS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total number of products
#   - Number of wireless products
#   - Number of premium products
#   - Most common category
cat("Total Products:", nrow(products_clean), "\n")
cat("Wireless Products:", sum(products_clean$is_wireless), "\n")
cat("Premium Products:", sum(products_clean$is_premium), "\n")
cat("Most Common Category:", 
    products_clean %>% 
    count(Category) %>% 
    slice_max(n, n = 1) %>% 
    pull(Category), "\n")
# Customer Sentiment
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total feedback entries
#   - Average sentiment score
#   - Percentage of positive reviews
#   - Percentage of negative reviews
cat("Total Feedback Entries:", nrow(feedback_clean), "\n")
cat("Average Sentiment Score:", round(mean(feedback_clean$sentiment_score, na.rm = TRUE), 2), "\n")
cat("Positive Reviews:", 
    round(sum(feedback_clean$sentiment_score > 0) / nrow(feedback_clean) * 100, 1), "%\n")
cat("Negative Reviews:", 
    round(sum(feedback_clean$sentiment_score < 0) / nrow(feedback_clean) * 100, 1), "%\n")

# Transaction Patterns
cat("\n📊 TRANSACTION PATTERNS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total transactions
#   - Date range (earliest to latest)
#   - Busiest weekday
#   - Weekend transaction percentage
cat("Total Transactions:", nrow(transactions_clean), "\n")
cat("Date Range:", 
    format(min(transactions_clean$date_parsed, na.rm = TRUE), "%b %d, %Y"), "to",
    format(max(transactions_clean$date_parsed, na.rm = TRUE), "%b %d, %Y"), "\n")
cat("Busiest Weekday:", as.character(weekday_patterns$trans_weekday[1]), "\n")
cat("Weekend Transactions:", 
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")
# Customer Recency
cat("\n👥 CUSTOMER RECENCY\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Number of recent customers (< 30 days)
#   - Number of at-risk customers (> 90 days)
#   - Percentage needing re-engagement
cat("Recent Customers (< 30 days):", sum(transactions_clean$days_since <= 30), "\n")
cat("At-Risk Customers (> 90 days):", sum(transactions_clean$days_since > 90), "\n")
cat("Needing Re-engagement:", 
    round(sum(transactions_clean$days_since > 90) / nrow(transactions_clean) * 100, 1), "%\n")
cat("\n", rep("=", 60), "\n")


 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
BUSINESS INTELLIGENCE SUMMARY
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

📦 PRODUCT ANALYSIS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Products: 75 
Wireless Products: 17 
Premium Products: 13 
Most Common Category: Electronics 

💬 CUSTOMER SENTIMENT
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Feedback Entries: 100 
Average Sentiment Score: 0.28 
Positive Reviews: 49 %
Negative Reviews: 29 %

📊 TRANSACTION PATTERNS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Transactions: 313 
Date Range: Mar 15, 2024 to Apr 05, 2024 
Busiest Weekday: Friday 
Weekend Transactions: 0 %

👥 CUSTOMER RECENCY
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Recent Customers (< 30 days): NA 


At-Risk Customers (> 90 days): NA 
Needing Re-engagement: NA %

 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 


In [46]:
# Task 7.2: Identify Top Products by Category
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
  count(Category, name = "product_count") %>% 
  arrange(desc(product_count)) %>%
  slice_head(n = 5)                                

cat("Top Product Categories:\n")
print(top_categories)
  

cat("Top Product Categories:\n")
print(top_categories)

Top Product Categories:
     Category product_count
1 Electronics            21
2   Computers            15
3       Audio            14
4          TV            14
5       Shoes            11
Top Product Categories:
     Category product_count
1 Electronics            21
2   Computers            15
3       Audio            14
4          TV            14
5       Shoes            11


## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

It made it easier to group and filter data accurately, for example, standardizing product categories allowed me to correctly count products in each category without duplicates caused by case differences or extra spaces, making it easier to analyze and get insights from. Another example is cleaning customer feedback text helped in accurately detecting sentiment words, leading to a more reliable sentiment analysis and from there understanding customer satisfaction levels by grouping them into "buckets" of positive and negative feedback, only thing I would've done different is to add a neutral category.



### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

I gathered that none of the products were gaming related, while a good portion were wireless and premium. A business could use this information to tailor their marketing strategies, focusing on promoting wireless and premium products more heavily, while also considering expanding their product line to include gaming products to capture that market.



### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**

Your answer here:

1. To identify sell patterns . For example, if a certain product sells more on weekends, the business can stock up inventory accordingly.
2. To identify peak business hours and days for staffing purposes, ensuring adequate customer service during high-traffic periods.
3. To identify seasonal trends in sales, allowing for targeted marketing campaigns during high-sales months. An example of this would be an increase in sales during december comes to show that people tend to buy more gifts and products during the holiday season.


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

Your answer here: Well since the data is outdated and the dates are showed from last year everything was showed at risk. But assuming this wasnt the case, I would prioritize re-engagement campaigns for "Moderate" customers first, this because moderate customers have shown interest "not too long ago" and are more likely to respond positively to outreach efforts. Next, I would maintain engagement with "Recent" customers through loyalty programs and personalized communications to keep them satisfied and encourage repeat business. Lastly, I would focus on "At Risk" customers with special offers or incentives to encourage them to return, as they have not transacted in a long time and may need more motivation, but these are potentially clients that are simply not interested anymore so not worth it investing too many resources there if they are simply not interested anymore.


### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Your answer here: The biggest set back or limitation I found to this was that there was not a category for neutral feedback, this could lead to misclassifying feedback that is neither positive nor negative, potentially skewing the analysis. Additionally, the approach may not capture the full context of customer sentiments, such as sarcasm or mixed feelings within a single review (For example, good product, not life changing). To improve products or customer service, businesses could use sentiment analysis to identify common pain points mentioned in negative reviews and address them directly, while also highlighting strengths from positive feedback to enhance marketing efforts. However, relying solely on keyword detection may overlook nuanced customer opinions, so a more sophisticated natural language processing approach could give better insights.



### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

Your answer here: I found string combining and manipulation particularly useful for retail stores that have both in person and online streams of sale. By combining string manipulation to clean and standardize product names and categories from both sales channels, and date analysis to track sales trends over time, a business could discover insights such as which products are most popular in each channel during specific times of the year. This could help optimize inventory management, marketing strategies, and promotional campaigns to boost sales and improve customer satisfaction across both channels.



## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**